src/crimson/common/gated.h

```diff
@@ -52,4 +68,52 @@ class Gated {
   seastar::gate pending_dispatch;
 };
 
-}// namespace crimson::common
+// gate_per_shard is a class that provides a gate for each shard.
+// It was introduced to provide a way to have gate for each shard
+// in a seastar application since gates are not supposed to be shared
+// across shards. ( https://tracker.ceph.com/issues/64332 )
+class gate_per_shard {
+ public:
+  gate_per_shard() : gates(seastar::smp::count) {}
+  explicit gate_per_shard(size_t shard_count) : gates(shard_count) {}
+  gate_per_shard(const gate_per_shard&) = delete;
+  gate_per_shard& operator=(const gate_per_shard&) = delete;
+  gate_per_shard(gate_per_shard&&) = default;
+  gate_per_shard& operator=(gate_per_shard&&) = default;
+  ~gate_per_shard() = default;
+
+  template <typename Func, typename T>
+  inline void dispatch_in_background(const char* what, T& who, Func&& func) {
+    (void) dispatch(what, who, std::forward<Func>(func));
+  }
+
+  template <typename Func, typename T>
+  inline auto dispatch(const char* what, T& who, Func&& func) {
+    return gates[seastar::this_shard_id()].dispatch(what, who, std::forward<Func>(func));
+  }
+
+  template <typename Func>
+  auto simple_dispatch(const char* what, Func&& func) {
+    return gates[seastar::this_shard_id()].simple_dispatch(what, std::forward<Func>(func));
+  }
+
+  bool is_closed() const {
+    return gates[seastar::this_shard_id()].is_closed();
+  }
+
+  seastar::future<> close_all() {
+    ceph_assert(gates.size() == seastar::smp::count);
+    std::vector<seastar::future<>> futures;
+    for (unsigned shard = 0; shard < seastar::smp::count; ++shard) {
+      futures.push_back(seastar::smp::submit_to(shard, [this, shard] {
+        return gates[shard].close();
+      }));
+    }
+    return seastar::when_all_succeed(futures.begin(), futures.end()).discard_result();
+  }
```

Comment:
https://github.com/ceph/ceph/pull/58986#discussion_r1718267126

Reply:
@Matan-B please review the parallel_for_each

Reply:
@Matan-B if you have some time to look